ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,338 @@
+          in
+          let _, _ =
+            Forkhelpers.execute_command_get_output !Xapi_globs.udevadm
+              ["settle"; "-t"; "30"]
+          in
+          ()
+      with Forkhelpers.Spawn_internal_error (err, _, _) ->
+        internal_error "Failed to reload driver selection (%s): %s" __FUNCTION__
+          err
+  )
+
+(* Versions are compared in segments separated by punctuation characters.
+   If both segments are integers, integer comparison is used, if at least
+   one of the segments is not an integer, lexicographic comparison is used.
+   Any segment compares larger against an empty segment.
+   Thus, "1.2.3" > "1.2";
+         "1.2.3" > "1.2."
+         "1.2.3" < "1.2.fix"
+         "1.2.3" < "1.2.3fix"
+*)
+module Version = struct
+  let compare a b =
+    let split str = Str.(split_delim (regexp "[.,:;+_-]") str) in
+
+    List.compare
+      (fun a b ->
+        Stdlib.(
+          match (int_of_string_opt a, int_of_string_opt b) with
+          | Some a, Some b ->
+              Int.compare a b
+          | _, _ ->
+              String.compare a b
+        )
+      )
+      (split a) (split b)
+end
+
+let get_driver_versions driver_name selected_dir all_versions_dir =
+  let selected_version =
+    try
+      let driver_realpath =
+        Unix.readlink (selected_dir // (driver_name ^ ".ko"))
+      in
+      Filename.(basename (dirname driver_realpath))
+    with _ ->
+      D.debug "Driver '%s' has no selected version" driver_name ;
+      ""
+  in
+  let active_version =
+    try
+      Result.value ~default:""
+        (Xapi_host_driver_helpers.get_version
+           ("/sys/module" // driver_name // "notes/note.xenserver")
+        )
+    with _ ->
+      D.debug "Driver '%s' has no active version" driver_name ;
+      ""
+  in
+  let driver_versions =
+    let arr = Sys.readdir (all_versions_dir // driver_name) in
+    Array.sort Version.compare arr ;
+    Array.to_list arr
+  in
+  (selected_version, active_version, driver_versions)
+
+let update_drivers_db ~__context ~host ~driver_paths db_drivers
+    filesystem_drivers =
+  try
+    let db_drivers_set, db_drivers_map =
+      db_drivers
+      |> List.fold_left
+           (fun (set, map) (ref, driver) ->
+             ( DriverSet.add driver.API.host_driver_name set
+             , DriverMap.add driver.API.host_driver_name (ref, driver) map
+             )
+           )
+           (DriverSet.empty, DriverMap.empty)
+    in
+    let filesystem_drivers =
+      Array.fold_left
+        (fun set name -> DriverSet.add name set)
+        DriverSet.empty filesystem_drivers
+    in
+    (* Modify drivers that were in the DB before and are still on the
+       filesystem in-place to keep references from outside intact *)
+    let updated_drivers =
+      DriverSet.inter db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.filter_map (fun driver_name ->
+             let self, db_driver = DriverMap.find driver_name db_drivers_map in
```

Comment:
If `DriverMap.find` is used here anyway, then the `DriverSet.inter` seems not necessary. The logic would be: for each driver in `filesystem_drivers`, `DriverMap.find_opt driver_name db_drivers_map |> Option.map (fun (ref, record) -> ...)`.